### PR TITLE
🤖 fix: fence stale compaction publications after history deletion

### PR DESCRIPTION
**Pull request by ThomasK33**
Deleting current provider rows or their boundary can leave a foreign compactor's captured generation valid, allowing stale initial, fallback, or boundary publication after cleanup. This change fences the removed occurrences before rewriting history, using the provider scanner's raw-aware suffix rules.

It covers single deletion, batch deletion, archive fallback, and the private deletion used by commitPartial. Deleting a readable separator can join malformed fragments into new reset evidence; the shared scanner probe detects that change while preserving the raw bytes. Readable user/system reset floors participate in deletion classification without changing provider reads. Oversized reset evidence stays raw, and protected active identities prevent deletion or truncation from falling back to an archived duplicate. Exact-limit valid rows remain addressable because the size check excludes the trailing LF. Retained boundaries and separators, existing reset evidence, incomplete tokens, and ineffective cuts preserve generation. A later history-write failure retains the conservative fence.

Protected parsed rows retain identity and sequence metadata separately from transformable rows. Partial rewrites and rename keep their cached counter above retained protected sequences. Seven counter regressions were reproduced; actual sequence reuse was not reproduced because ordinary appends already refresh durable history. Reasoning-only deletions inherit the conservative provider-context classification from merged #4135. The pending-state prerequisite #4138 has also merged; this PR is the remaining deletion layer. Runtime activation is separate.

Validation: 518 affected tests and full static checks pass on the main-integrated candidate. Coverage includes raw-fragment joining, malformed-role reset floors, protected active/archive duplicate IDs, exact-limit and oversized rows, sequence accounting, real provider history, and stale initial/fallback/boundary publication. The exact-rollback fixture restores captured history bytes so its unchanged-generation assertions remain meaningful after generic deletion gains fencing.

Edit/fork truncation now refuses a protected active target before archive fallback when no readable active match exists. Four regressions reproduced destructive fallback to an archived duplicate; twelve cases cover oversized and ambiguous reset rows in both keep/remove modes. Refusal preserves both files byte-for-byte, generation, and the publication journal. Readable active duplicates retain precedence, while unrelated protected IDs still allow legitimate archive truncation. The correction reuses existing identity metadata and the single-deletion guard; independent review approved it.

The queued-base CI failure was isolated to four fault-injection cases after #4146 changed generation publication to write a staging sibling and then rename it. The fixture now matches that exact generation artifact's staging prefix, keeps history-stage matching exact, and asserts that injection fired. All four regressions, all eight write-failure controls, and the 349-test history suite pass on the previously failing merge-group base. Production behavior is unchanged by this integration correction. Nix formatting was skipped because Nix is unavailable.

Linux E2E also now runs with one worker because all Electron workers share one Xvfb display and clipboard. On the exact failing queue candidate, native probes reproduced three clipboard denials in four competing-window runs: two lost focus before validation, and one during the awaited isolated-world activation check. Focusing the parent before clicking a self-closing auth popup reproduced the same closed-target error on attempt nine; authentication completed, the remote stayed connected, and the local renderer and projects remained intact. Security checks and all original tests remain unchanged.

The unmodified remote-connection file passes all three tests with one Linux worker (34.9 seconds); 30 focused popup controls also pass. Reproduction used Linux arm64 with pinned Electron 40.9.3, Playwright 1.57.0, and Bun 1.3.5; the failing CI runner was x64. Full static checks and both workflow linters pass for the final CI-only correction. The worker-count repair was verified at b82705b1f by the full Linux x64 CI suite: 55 tests passed and 21 skipped using one worker, including all remote-connection tests. Measured test execution took 8.5 minutes; the complete job took 9m56s including setup and build ([CI result](https://github.com/coder/xum/actions/runs/34279578520/job/102240858941)). Serial execution trades some throughput for stable native focus without reducing coverage.

Regression risk is in classifying which removed occurrences affect provider context while preserving malformed privacy floors and occupied sequences. Tests cover retained-floor controls, duplicate IDs, restart, and generation/history-write failure ordering using real HistoryService storage.

---

_Generated with `xum` • Model: `unavailable` • Thinking: `unavailable` • Cost: `$unavailable`_

### CODE REVIEW

**Comment by ThomasK33** [comment]
@codex review

Please review this bounded deletion-fencing follow-up independently of #4135's reset/truncation layer. It addresses that PR's finding 3958805613 through the single, batch, archive, and private partial-cleanup deletion paths. Focus on actual removed-occurrence classification, raw reset preservation, and generation/write failure ordering. The production change is 51 lines; independent review approved.

---

_Generated with `xum` • Model: `unavailable` • Thinking: `unavailable` • Cost: `$unavailable`_

**Comment by ThomasK33** [comment]
Addressed finding 3959648277: deleting a readable separator can join raw fragments into a new reset floor. The deletion fence now checks newly exposed reset evidence with the same incremental probe used by provider-history scanning. Raw bytes remain intact, and retained separators/boundaries, existing evidence, and incomplete tokens preserve generation.

Nine new regressions reproduced the reported behavior before the fix. All 302 affected tests and full static checks pass; independent review approved. This PR is now the fixed top of phase stack #4135 → #4138 → #4148, with no runtime activation added.

@codex review

---

_Generated with `xum` • Model: `unavailable` • Thinking: `unavailable` • Cost: `$unavailable`_

**Comment by ThomasK33** [comment]
Addressed both findings:

- 3960203163: rewrite classification now preserves oversized raw reset evidence just as the provider scanner does. Deletion is refused and bytes remain intact. Size checks exclude the trailing LF: a valid row at the exact limit remains deletable and advances generation; a row one byte above the limit remains preserved.
- 3960203170: reasoning-only context is now conservatively fenced by the shared helper in parent #4135, including all its mutation families. This PR retains single/batch/archive deletion regressions and inherits that implementation without duplicating it.

Integrated F1 and G1's final reviewed fixes. All 395 affected tests and full static checks pass; independent review approved the classification and exact-limit correction. The own diff contains only history scanner/service/tests, and runtime activation remains separate.

@codex review

---

_Generated with `xum` • Model: `unavailable` • Thinking: `unavailable` • Cost: `$unavailable`_

**Comment by ThomasK33** [comment]
Integrated the validated seven-line directory-healing correction from #4138; all four existing deletion-layer commits remain identical by range-diff.

CI exposed an inherited fixture using generic boundary deletion while expecting generation to stay unchanged. This layer intentionally fences that deletion. The test now restores captured real history bytes to model an already-completed exact rollback, preserving all six unchanged-generation/boundary cases and their restart/attachment assertions. No production behavior changed in this integration correction.

All 454 combined tests and full static checks pass. Independent review approved the one-file test adjustment. This remains the fixed top of #4135 → #4138 → #4148; the phase will be submitted together after every member is ready.

@codex review

---

_Generated with `xum` • Model: `unavailable` • Thinking: `unavailable` • Cost: `$unavailable`_

**Comment by ThomasK33** [comment]
Addressed both outstanding deletion findings:

- 3960656026: deletion classification includes readable user/system manual-reset floors that provider requests exclude. Removing those floors advances the generation even with an empty tail. Provider-read defaults remain unchanged; newer retained floors still keep sealed removals neutral.
- 3960656031: protected raw rows retain their parsed identity for active-first lookup. A protected-only active match refuses deletion rather than removing an archived duplicate. Batch refusal and readable active duplicate behavior remain covered.

All 482 affected tests and full static checks pass; independent review approved. The fix stays within the existing scanner, history service, and behavioral tests. No activation or new parser framework was added.

@codex review

---

_Generated with `xum` • Model: `unavailable` • Thinking: `unavailable` • Cost: `$unavailable`_

**Comment by ThomasK33** [comment]
@codex review

Integrated current main so this phase includes merged CI fix #4149. GitHub confirmed the previous top-layer unit job exceeded its old 15-minute limit; the merged workflow supplies the corrected limit and comment gate. Every own-layer commit remains identical by git range-diff.

Validation on the integrated base: 374 fencing tests, 63 pending-protocol tests, and 482 combined top-layer tests pass; full static checks pass at the phase top. No source changes were needed during integration.

The complete phase remains #4135 → #4138 → #4148. Please confirm the current head; all three will be queued together after refreshed review and CI pass.

---

_Generated with `xum` • Model: `unavailable` • Thinking: `unavailable` • Cost: `$unavailable`_

**Comment by ThomasK33** [comment]
Addressed finding 3961198654 at 96f88ea554. Protected raw rows now retain parsed identity and sequence metadata independently of transformable messages. Deletion, truncation, percentage cuts, and workspace rename preserve their occupied sequence floor; full history clear still resets it.

Seven cached-counter regressions were reproduced before this fix. Actual sequence reuse was not reproduced because ordinary appends already refresh from durable history. The correction stays within the raw-row preservation invariant of this PR: 42 changed production lines and 18 regression cases.

All 506 affected tests, including inherited G1 coverage, and full static checks passed on the final integrated head. Independent review approved. Nix formatting was skipped because Nix is unavailable.

@codex review

---

_Generated with `xum` • Model: `unavailable` • Thinking: `unavailable` • Cost: `$unavailable`_

**Comment by ThomasK33** [comment]
@codex review

Fixed the four generation-write fault-injection failures from merge-group run 34270828968. The generation writer introduced by merged #4146 writes an artifact-specific staging sibling before rename; the fixture previously matched only the final path and never injected the failure. It now matches the full generation staging prefix, preserves exact history-path matching, and asserts that the intended injection fired. All history, counter, generation, journal, and partial-state assertions remain intact. This correction changes only the test fixture.

Reproduced all four failures on the untouched queued base, then verified four passing generation cases, all eight write-failure controls, and all 349 history tests after the correction. On actual main with merged #4135 and #4138, all 506 combined tests and full static validation pass. GitHub's automatic restack preserves the seven existing F2 commits; range-diff shows only an import-context variation from main. No production changes were needed during integration.

Please review the current candidate. The coordinator owns queueing and merge decisions.

---

_Generated with `xum` • Model: `unavailable` • Thinking: `unavailable` • Cost: `$unavailable`_

**Comment by ThomasK33** [comment]
@codex review

Added the approved CI-only repair for Linux E2E failures in the queued candidate. Electron workers shared one Xvfb display and clipboard; native probes reproduced focus loss before and during clipboard permission checks (three failures in four competing-window runs). Focusing the parent before a self-closing auth popup reproduced the same closed-target click error, while authentication completed and the remote connection, local renderer identity, and projects remained healthy.

Linux E2E now uses one worker, with a rationale comment. Every existing test and production security check remains unchanged. The unmodified remote-connection file passes 3/3 tests with one Linux worker, and 30 focused popup controls pass. Full static validation and actionlint/zizmor pass. Reproduction was Linux arm64; CI is x64. Roughly 15 minutes from summed previous CI attempt durations is only an estimate; this CI run will measure full serial runtime.

This commit changes only the workflow invocation and comment; the reviewed F2 implementation and earlier fault-injection correction are unchanged. The coordinator owns queueing and merge decisions.

---

_Generated with `xum` • Model: `unavailable` • Thinking: `unavailable` • Cost: `$unavailable`_

File: src/node/services/historyService.ts [hunk 1]
```diff
@@ -2621,20 +2662,39 @@ export class HistoryService {
...
         raw: line,
-        message: this.parseMessages(text, filePath, (value) =>
-          isReadableHistoryMessage(value) &&
-          !(hasRawResetMarker(text) && hasAmbiguousResetKeys(text))
-            ? normalizeLegacyMuxMetadata(value)
-            : null
-        )[0],
+        message: protectedReset ? undefined : parsed,
+        // Keep identity and sequence accounting even when the raw floor cannot be rewritten.
+        protectedMessage: protectedReset ? parsed : undefined,
```
[the remark this answers is not in the document]
Comment (ThomasK33) [reply src/node/services/historyService.ts:2682]: Addressed finding 3962413670. `truncateAfterMessage` now applies the existing protected-active refusal inside the no-readable-match branch, before archive fallback. A protected active ID therefore cannot redirect an edit or fork to an archived duplicate. Readable active matches retain precedence, and unrelated protected IDs still permit legitimate archive truncation.

The four protected-target regressions failed before the guard; eight positive controls already passed. All twelve now pass, asserting exact active/archive bytes and unchanged generation/journal on refusal. Both oversized and ambiguous rows and both keep/remove modes are covered. All 518 affected tests and full static checks pass; independent review approved the bounded correction. The Linux CI worker-count repair is unchanged.

---

_Generated with `xum` • Model: `unavailable` • Thinking: `unavailable` • Cost: `$unavailable`_

**Comment by ThomasK33** [comment]
@codex review

Addressed and resolved finding 3962413670 with the narrow protected-active guard before truncation's archive fallback. The original four failing cases now pass alongside readable-active and unrelated-protected-ID archive controls. Twelve regression cases cover oversized/ambiguous rows, edit/fork target modes, exact file preservation, and unchanged generation/publication state on refusal.

All 518 affected tests and full static checks pass. Independent review approved the six-line production guard and regression matrix. The existing Linux CI serialization repair and every production security check remain intact. Please review the current head; the coordinator owns queueing and merge decisions.

---

_Generated with `xum` • Model: `unavailable` • Thinking: `unavailable` • Cost: `$unavailable`_

**Comment by ThomasK33** [comment]
@codex review

Please provide an explicit approval or actionable findings for the unchanged head `71e8e36221d785db6aa41cc66f5ed08e5ca4d988`. The previous code review completed at 21:55:15 UTC and security review completed at 21:57:54 UTC, but no approval comment or thumbs-up appeared. A completed status alone does not satisfy this repository's approval gate.

All current-head test and build jobs pass, including Linux E2E with 55 passing tests and 21 skips using one worker. All seven review threads are resolved, and security review found no issues. No source changes were made after the prior review. Please give a clear current-head verdict; the coordinator will retry the stale CI comments gate after review finishes and owns queueing and merge decisions.

---

_Generated with `xum` • Model: `unavailable` • Thinking: `unavailable` • Cost: `$unavailable`_
